platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/CapabilitiesRules.java

```diff
@@ -36,10 +36,10 @@ public void addCapabilitiesAction(VariantMetadataRules.VariantAction<? super Mut
         actions.add(action);
     }
 
-    public CapabilitiesMetadata execute(VariantResolveMetadata variant, MutableCapabilitiesMetadata capabilities) {
+    public ImmutableCapabilities execute(VariantResolveMetadata variant, MutableCapabilitiesMetadata capabilities) {
         for (VariantMetadataRules.VariantAction<? super MutableCapabilitiesMetadata> action : actions) {
             action.maybeExecute(variant, capabilities);
         }
-        return capabilities.asImmutable();
+        return ImmutableCapabilities.of(capabilities.asImmutable().getCapabilities());
```

Comment:
Shouldn't the return type of `asImmutable` be adjusted here so there's no need to re-wrap it? It already makes an `ImmutableCapabilities`.

Reply:
`asImmutable` is a public API method and does not return an `ImmutableCapabiliites`.  
However I adjusted this to avoid as much re-wrapping as possible.